# Orders_df - by Hong Eng

### Order Dataset
This is the core dataset. From each order you might find all other information.

## Extract

In [1]:
import pandas as pd

orders_df = pd.read_csv("C:/Users/Kelley/Desktop/Project/Final Project/data/olist_orders_dataset.csv")

orders_df.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06  2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39  2018-02-13 22:20:29   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   

  order_estimated_delivery_date  
0           2017-10-18 00:00:00  
1           2018-08-13 00:00:00  
2           2018-09-04 00:00:00  
3           2017-12-15 00:00:00  
4           2018-02-26 00:00:00

## Explore Dataset

In [2]:
# Exploring the orders dataset.
orders_df.info()
orders_df.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   order_id                       99441 non-null  object
 1   customer_id                    99441 non-null  object
 2   order_status                   99441 non-null  object
 3   order_purchase_timestamp       99441 non-null  object
 4   order_approved_at              99281 non-null  object
 5   order_delivered_carrier_date   97658 non-null  object
 6   order_delivered_customer_date  96476 non-null  object
 7   order_estimated_delivery_date  99441 non-null  object
dtypes: object(8)
memory usage: 6.1+ MB


order_id                       customer_id  \
count                              99441                             99441   
unique                             99441                             99441   
top     e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
freq                                   1                                 1   

       order_status order_purchase_timestamp    order_approved_at  \
count         99441                    99441                99281   
unique            8                    98875                90733   
top       delivered      2018-04-11 10:48:14  2018-02-27 04:31:10   
freq          96478                        3                    9   

       order_delivered_carrier_date order_delivered_customer_date  \
count                         97658                         96476   
unique                        81018                         95664   
top             2018-05-09 15:48:00           2018-05-08 23:38:46   
freq                             47                             3   

       order_estimated_delivery_date  
count                          99441  
unique                           459  
top              2017-12-20 00:00:00  
freq                             522

## Columns

1. <span style="color: blue;">**order_id**</span> --- <span style="color: red;">*(Primary Key)*</span>
    - order unique identifier

2. <span style="color: blue;">**customer_id**</span> --- <span style="color: red;">*(Foreign Key)*</span>
    - Key to the customer dataset. Each order has a unique customer_id.
    
3. <span style="color: blue;">**order_status**</span>
    - Reference to the order status (delivered, shipped, etc).

4. <span style="color: blue;">**order_purchase_timestamp**</span>
    - Shows the purchase timestamp.

5. <span style="color: blue;">**order_approved_at**</span>
    - Shows the payment approval timestamp.

6. <span style="color: blue;">**order_delivered_carrier_date**</span>
    - Shows the order posting timestamp. When it was handled to the logistic partner.

7. <span style="color: blue;">**order_delivered_customer_date**</span>
    - Shows the actual order delivery date to the customer.

8. <span style="color: blue;">**order_estimated_delivery_date**</span>
    - Shows the estimated delivery date that was informed to customer at the purchase moment. 

### count of order_id = unique of order_id, meaning there is no duplicate orders

In [3]:
# Returning unique values for order status.
orders_df["order_status"].unique()


array(['delivered', 'invoiced', 'shipped', 'processing', 'unavailable',
       'canceled', 'created', 'approved'], dtype=object)

## Transform

#### data type to change

 0   order_id                       99441 non-null  object  No change  

 1   customer_id                    99441 non-null  object  No change
 
 2   order_status                   99441 non-null  object  to  **category**
 
 3   order_purchase_timestamp       99441 non-null  object  to  **datetime**
 
 4   order_approved_at              99281 non-null  object  to  **datetime**
 
 5   order_delivered_carrier_date   97658 non-null  object  to  **datetime**
 
 6   order_delivered_customer_date  96476 non-null  object  to  **datetime**
 
 7   order_estimated_delivery_date  99441 non-null  object  to  **datetime**

#### Convert the order_purchase_timestamp, order_approved_at, order_delivered_carrier_date, order_delivered_customer_date, order_estimated_delivery_date to **datetime** data type

In [4]:
# Converting order_purchase_timestamp, order_approved_at, order_delivered_carrier_date, order_delivered_customer_date, order_estimated_delivery_date
orders_df['order_purchase_timestamp'] = pd.to_datetime(orders_df['order_purchase_timestamp'])
orders_df['order_approved_at'] = pd.to_datetime(orders_df['order_approved_at'])
orders_df['order_delivered_carrier_date'] = pd.to_datetime(orders_df['order_delivered_carrier_date'])
orders_df['order_delivered_customer_date'] = pd.to_datetime(orders_df['order_delivered_customer_date'])
orders_df['order_estimated_delivery_date'] = pd.to_datetime(orders_df['order_estimated_delivery_date'])

# Checking datatype after changing
orders_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype         
---  ------                         --------------  -----         
 0   order_id                       99441 non-null  object        
 1   customer_id                    99441 non-null  object        
 2   order_status                   99441 non-null  object        
 3   order_purchase_timestamp       99441 non-null  datetime64[ns]
 4   order_approved_at              99281 non-null  datetime64[ns]
 5   order_delivered_carrier_date   97658 non-null  datetime64[ns]
 6   order_delivered_customer_date  96476 non-null  datetime64[ns]
 7   order_estimated_delivery_date  99441 non-null  datetime64[ns]
dtypes: datetime64[ns](5), object(3)
memory usage: 6.1+ MB


### Replace wrong spelling of 'canceled' to 'cancelled'

In [5]:
# Replacing wrong spelling 'canceled' to 'cancelled'
orders_df["order_status"] = orders_df["order_status"].replace("canceled", "cancelled")

#### Convert the order_status to **category** data type.
This will help optimize memory usage and improve performance when working with this column.

In [6]:
orders_df["order_status"] = orders_df["order_status"].astype("category")

# Checking datatype after changing
orders_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype         
---  ------                         --------------  -----         
 0   order_id                       99441 non-null  object        
 1   customer_id                    99441 non-null  object        
 2   order_status                   99441 non-null  category      
 3   order_purchase_timestamp       99441 non-null  datetime64[ns]
 4   order_approved_at              99281 non-null  datetime64[ns]
 5   order_delivered_carrier_date   97658 non-null  datetime64[ns]
 6   order_delivered_customer_date  96476 non-null  datetime64[ns]
 7   order_estimated_delivery_date  99441 non-null  datetime64[ns]
dtypes: category(1), datetime64[ns](5), object(2)
memory usage: 5.4+ MB


the values of the datetime columns are reflection of an instance when the data was collected for the dataset. 
And therefore, 
no. of order approved is not equal no. delivered_carrier and 
no. order approves is not equal to no. delivered-customer
??????????????

### Exploring relationship of null values Vs order_status

In [7]:
# Count null values in each column
null_counts = orders_df.isnull().sum()

# Display the result
print(null_counts)

order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                 160
order_delivered_carrier_date     1783
order_delivered_customer_date    2965
order_estimated_delivery_date       0
dtype: int64


### Exploring nulls in 'order_approved_at'

In [8]:
# Subsetting with order_approved_at nulls
approved_null = orders_df[orders_df['order_approved_at'].isnull()]

approved_null.head()

order_id                       customer_id  \
1130  00b1cb0320190ca0daa2c88b35206009  3532ba38a3fd242259a514ac2b6ae6b6   
1801  ed3efbd3a87bea76c2812c66a0b32219  191984a8ba4cbb2145acb4fe35b69664   
1868  df8282afe61008dc26c6c31011474d02  aa797b187b5466bc6925aaaa4bb3bed1   
2029  8d4c637f1accf7a88a4555f02741e606  b1dd715db389a2077f43174e7a675d07   
2161  7a9d4c7f9b068337875b95465330f2fc  7f71ae48074c0cfec9195f88fcbfac55   

     order_status order_purchase_timestamp order_approved_at  \
1130    cancelled      2018-08-28 15:26:39               NaT   
1801    cancelled      2018-09-20 13:54:16               NaT   
1868    cancelled      2017-03-04 12:14:30               NaT   
2029    cancelled      2018-08-29 16:27:49               NaT   
2161    cancelled      2017-05-01 16:12:39               NaT   

     order_delivered_carrier_date order_delivered_customer_date  \
1130                          NaT                           NaT   
1801                          NaT                           NaT   
1868                          NaT                           NaT   
2029                          NaT                           NaT   
2161                          NaT                           NaT   

     order_estimated_delivery_date  
1130                    2018-09-12  
1801                    2018-10-17  
1868                    2017-04-10  
2029                    2018-09-13  
2161                    2017-05-30

In [9]:
approved_null.shape

(160, 8)

In [10]:
# Unique row count for subset
unique_row_count_approved_null = approved_null.nunique()

print(unique_row_count_approved_null)

order_id                         160
customer_id                      160
order_status                       3
order_purchase_timestamp         160
order_approved_at                  0
order_delivered_carrier_date      13
order_delivered_customer_date     14
order_estimated_delivery_date    109
dtype: int64


In [11]:
# Returning unique values for order status for approved_null df
approved_null["order_status"].unique()

['cancelled', 'delivered', 'created']
Categories (8, object): ['approved', 'cancelled', 'created', 'delivered', 'invoiced', 'processing', 'shipped', 'unavailable']

In [12]:
# Count null values in each column
null_counts_approved_null = approved_null.isnull().sum()

# Display the result
print(null_counts_approved_null)

order_id                           0
customer_id                        0
order_status                       0
order_purchase_timestamp           0
order_approved_at                160
order_delivered_carrier_date     146
order_delivered_customer_date    146
order_estimated_delivery_date      0
dtype: int64


Subset approved_null with status delivered to customer.

In [13]:
# Subset approved_null with status delivered to customer
subset_approved_null_delivered_cust = approved_null.dropna(subset=["order_delivered_customer_date"])

subset_approved_null_delivered_cust

order_id                       customer_id  \
5323   e04abd8149ef81b95221e88f6ed9ab6a  2127dc6603ac33544953ef05ec155771   
16567  8a9adc69528e1001fc68dd0aaebbb54a  4c1ccc74e00993733742a3c786dc3c1f   
19031  7013bcfc1c97fe719a7b5e05e61c12db  2941af76d38100e0f8740a374f1a5dc3   
22663  5cf925b116421afa85ee25e99b4c34fb  29c35fc91fc13fb5073c8f30505d860d   
23156  12a95a3c06dbaec84bcfb0e2da5d228a  1e101e0daffaddce8159d25a8e53f2b2   
26800  c1d4211b3dae76144deccd6c74144a88  684cb238dc5b5d6366244e0e0776b450   
38290  d69e5d356402adc8cf17e08b5033acfb  68d081753ad4fe22fc4d410a9eb1ca01   
39334  d77031d6a3c8a52f019764e68f211c69  0bf35cac6cc7327065da879e2d90fae8   
48401  7002a78c79c519ac54022d4f8a65e6e8  d5de688c321096d15508faae67a27051   
61743  2eecb0d85f281280f79fa00f9cec1a95  a3d3c38e58b9d2dfb9207cab690b6310   
63052  51eb2eebd5d76a24625b31c33dd41449  07a2a7e0f63fd8cb757ed77d4245623c   
67697  88083e8f64d95b932164187484d90212  f67cd1a215aae2a1074638bbd35a223a   
72407  3c0b8706b065f9919d0505d3b3343881  d85919cb3c0529589c6fa617f5f43281   
84999  2babbb4b15e6d2dfe95e2de765c97bce  74bebaf46603f9340e3b50c6b086f992   

      order_status order_purchase_timestamp order_approved_at  \
5323     delivered      2017-02-18 14:40:00               NaT   
16567    delivered      2017-02-18 12:45:31               NaT   
19031    delivered      2017-02-18 13:29:47               NaT   
22663    delivered      2017-02-18 16:48:35               NaT   
23156    delivered      2017-02-17 13:05:55               NaT   
26800    delivered      2017-01-19 12:48:08               NaT   
38290    delivered      2017-02-19 01:28:47               NaT   
39334    delivered      2017-02-18 11:04:19               NaT   
48401    delivered      2017-01-19 22:26:59               NaT   
61743    delivered      2017-02-17 17:21:55               NaT   
63052    delivered      2017-02-18 15:52:27               NaT   
67697    delivered      2017-02-18 22:49:19               NaT   
72407    delivered      2017-02-17 15:53:27               NaT   
84999    delivered      2017-02-18 17:15:03               NaT   

      order_delivered_carrier_date order_delivered_customer_date  \
5323           2017-02-23 12:04:47           2017-03-01 13:25:33   
16567          2017-02-23 09:01:52           2017-03-02 10:05:06   
19031          2017-02-22 16:25:25           2017-03-01 08:07:38   
22663          2017-02-22 11:23:10           2017-03-09 07:28:47   
23156          2017-02-22 11:23:11           2017-03-02 11:09:19   
26800          2017-01-25 14:56:50           2017-01-30 18:16:01   
38290          2017-02-23 03:11:48           2017-03-02 03:41:58   
39334          2017-02-23 07:23:36           2017-03-02 16:15:23   
48401          2017-01-27 11:08:05           2017-02-06 14:22:19   
61743          2017-02-22 11:42:51           2017-03-03 12:16:03   
63052          2017-02-23 03:09:14           2017-03-07 13:57:47   
67697          2017-02-22 11:31:06           2017-03-02 12:06:06   
72407          2017-02-22 11:31:30           2017-03-03 11:47:47   
84999          2017-02-22 11:23:11           2017-03-03 18:43:43   

      order_estimated_delivery_date  
5323                     2017-03-17  
16567                    2017-03-21  
19031                    2017-03-17  
22663                    2017-03-31  
23156                    2017-03-20  
26800                    2017-03-01  
38290                    2017-03-27  
39334                    2017-03-22  
48401                    2017-03-16  
61743                    2017-03-20  
63052                    2017-03-29  
67697                    2017-03-21  
72407                    2017-03-23  
84999                    2017-03-31

No. of orders with order_approved_at==null = 160 ----- Of the 160, there were 160-146=14 orders delivered.

There are timestamps for both delivered_carrier and delivered_customer, it shows that orders were delivered.

Possible reasons: seller did not update approval / system did not register seller's approval
##### Conclusion - Retain records and null values.

### Exploring nulls in order_delivered_carrier_date

In [14]:
# Filter and count order_status where order_delivered_carrier_date is null.
filtered_status_counts = orders_df.loc[orders_df['order_delivered_carrier_date'].isnull(), 'order_status'].value_counts()

print("Order status counts where order_delivered_carrier_date is null:")
for status, count in filtered_status_counts.items():
    print(f"{status}: {count}")

# Calculate and print the total number of orders
total_orders = filtered_status_counts.sum()
print(f"\nTotal: {total_orders}")

Order status counts where order_delivered_carrier_date is null:
unavailable: 609
cancelled: 550
invoiced: 314
processing: 301
created: 5
approved: 2
delivered: 2
shipped: 0

Total: 1783


In [15]:
# subset order_delivered_carrier_date null and status delivered.
subset = orders_df.loc[(orders_df['order_delivered_carrier_date'].isnull()) & (orders_df['order_status'] == 'delivered')]

print("Subset of orders where order_delivered_carrier_date is null and order_status is 'delivered':")
subset


Subset of orders where order_delivered_carrier_date is null and order_status is 'delivered':


order_id                       customer_id  \
73222  2aa91108853cecb43c84a5dc5b277475  afeb16c7f46396c0ed54acb45ccaaa40   
92643  2d858f451373b04fb5c984a1cc2defaf  e08caf668d499a6d643dafd7c5cc498a   

      order_status order_purchase_timestamp   order_approved_at  \
73222    delivered      2017-09-29 08:52:58 2017-09-29 09:07:16   
92643    delivered      2017-05-25 23:22:43 2017-05-25 23:30:16   

      order_delivered_carrier_date order_delivered_customer_date  \
73222                          NaT           2017-11-20 19:44:47   
92643                          NaT                           NaT   

      order_estimated_delivery_date  
73222                    2017-11-14  
92643                    2017-06-23

#### Conclusion
##### (a) There is 1 record with status delivered carrier_date null but delivered_customer_date not null. - Retain record. For this case, we take order_delivered_customer_date as the determining factor that record should be valid.
##### (b) There is 1 record with order_status delivered but nulls for order_delivered_carrier_date and order_delivered_customer_date. - Create new order_status category for this row (b) - 'delivered with no date'
Update category of this row at - Exploring nulls in order_delivered_customer_date section below.
##### (c) Uneventful for other records with delivered_carrier_date nulls. They are due to various reasons as stated in order_status - Retain records.

Add new order_status category for 'delivered with no dates'

In [16]:
# Add the new category to the existing categories
orders_df['order_status'] = orders_df['order_status'].cat.add_categories(['delivered with no date'])

### Exploring nulls in order_delivered_customer_date

In [17]:
# Filter and count order_status where order_delivered_customer_date is null.
filtered_status_counts_dlvcust = orders_df.loc[orders_df['order_delivered_customer_date'].isnull(), 'order_status'].value_counts()

print("Order status counts where order_delivered_carrier_date is null:")
for status, count in filtered_status_counts_dlvcust.items():
    print(f"{status}: {count}")

# Calculate and print the total number of orders
total_orders_dlvcust = filtered_status_counts_dlvcust.sum()
print(f"\nTotal: {total_orders}")

Order status counts where order_delivered_carrier_date is null:
shipped: 1107
cancelled: 619
unavailable: 609
invoiced: 314
processing: 301
delivered: 8
created: 5
approved: 2
delivered with no date: 0

Total: 1783


##### 8 orders have status=delivered but delivered_customer_date is null. 
##### Change order_status for these 7 orders to 'delivered with no date'

In [18]:
# Update category for the specific rows
orders_df.loc[(orders_df['order_delivered_customer_date'].isnull()) & 
              (orders_df['order_status'] == 'delivered'), 'order_status'] = 'delivered with no date'

# Verify the changes

check_changestatus = orders_df.loc[
    (orders_df['order_delivered_customer_date'].isnull()) & 
    (orders_df['order_status'] == 'delivered')]

changedstatus = orders_df.loc[
    (orders_df['order_delivered_customer_date'].isnull()) & 
    (orders_df['order_status'] == 'delivered with no date')]

print("\033[4mOriginal:\033[0m")
print(check_changestatus)

print("\n\033[4mchanged status:\033[0m")
print(changedstatus.head)

# check number of records changed
changedstatus.shape

Original:
Empty DataFrame
Columns: [order_id, customer_id, order_status, order_purchase_timestamp, order_approved_at, order_delivered_carrier_date, order_delivered_customer_date, order_estimated_delivery_date]
Index: []

changed status:
<bound method NDFrame.head of                                order_id                       customer_id  \
3002   2d1e2d5bf4dc7227b3bfebb81328c15f  ec05a6d8558c6455f0cbbd8a420ad34f   
20618  f5dd62b788049ad9fc0526e3ad11a097  5e89028e024b381dc84a13a3570decb4   
43834  2ebdfc4f15f23b91474edf87475f108e  29f0540231702fda0cfdee0a310f11aa   
79263  e69f75a717d64fc5ecdfae42b2e8e086  cfda40ca8dd0a5d486a9635b611b398a   
82868  0d3268bad9b086af767785e3f0fc0133  4f1d63d35fb7c8999853b2699f5c7649   
92643  2d858f451373b04fb5c984a1cc2defaf  e08caf668d499a6d643dafd7c5cc498a   
97647  ab7c89dc1bf4a1ead9d6ec1ec8968a84  dd1b84a7286eb4524d52af4256c0ba24   
98038  20edc82cf5400ce95e1afacc25798b31  28c37425f1127d887d7337f284080a0f   

                 order_status order_pur

(8, 8)

### Explore cancelled orders with delivered_customer_date

In [19]:
# subset order_status cancelled where order_delivered_customer_date is not null
cancelled_dlvcust = orders_df[orders_df['order_delivered_customer_date'].notnull()]
status_counts_cancelled_dlvcust = cancelled_dlvcust['order_status'].value_counts()
print(status_counts_cancelled_dlvcust)

order_status
delivered                 96470
cancelled                     6
approved                      0
created                       0
invoiced                      0
processing                    0
shipped                       0
unavailable                   0
delivered with no date        0
Name: count, dtype: int64


#### There are 6 cancelled orders but have delivered to customer date.

### Conclusion: 
#### Possible reasons: 
- Orders may already be in transit when customer cancelled.
- Orders were delivered but cancelled and waiting for return or refund.
- Cancelled orders and delivery info can be for different analytics purposes. 
- Therefore retain records.


## Load
### Create csv file from orders_df

In [20]:
# Creating csv file from orders_df
try:
    orders_df.to_csv("C:/Users/Kelley/Desktop/Project/Final Project/cleaned_orders.csv", index=False, mode="w")
    print("CSV file successfully created.")
except Exception as e:
    print(f"Error creating CSV file: {e}")

CSV file successfully created.


### Load to postgresql

In [21]:
''' #load to postgreql

#%load_ext sql

# Replace the connection string with your actual PostgreSQL credentials
%sql postgresql://username:password@localhost:5432/mydatabase

# Write the DataFrame to the 'orders' table and set primary key
%sql DROP TABLE IF EXISTS orders;
%sql PERSIST orders_df;

%sql ALTER TABLE orders ADD PRIMARY KEY (order_id);

# Set 'customer_id' as a foreign key (assuming 'customers' is the parent table)
%sql ALTER TABLE orders ADD CONSTRAINT fk_customer FOREIGN KEY (customer_id) REFERENCES customers(customer_id);
''' 

" #load to postgreql\n\n#%load_ext sql\n\n# Replace the connection string with your actual PostgreSQL credentials\n%sql postgresql://username:password@localhost:5432/mydatabase\n\n# Write the DataFrame to the 'orders' table and set primary key\n%sql DROP TABLE IF EXISTS orders;\n%sql PERSIST orders_df;\n\n%sql ALTER TABLE orders ADD PRIMARY KEY (order_id);\n\n# Set 'customer_id' as a foreign key (assuming 'customers' is the parent table)\n%sql ALTER TABLE orders ADD CONSTRAINT fk_customer FOREIGN KEY (customer_id) REFERENCES customers(customer_id);\n"